vuu-ui/packages/vuu-layout/src/layout-persistence/LocalLayoutPersistenceManager.ts

```diff
@@ -8,72 +8,131 @@ const metadataSaveLocation = "layouts/metadata";
-    const layout = this.loadLayouts().filter(layout => layout.id === id);
-
-    switch (layout.length) {
-      case 1: {
-        return layout[0].json;
-      }
-      case 0: {
-        console.log(`WARNING: no layout exists for ID "${id}"; returning empty layout`);
-        return {} as LayoutJSON;
-      }
-      default: {
-        console.log(`WARNING: multiple layouts exist for ID "${id}"; returning first instance`)
-        return layout[0].json;
-      }
-    }
+  loadLayout(id: string): Promise<LayoutJSON> {
+    return new Promise((resolve, reject) => {
+      this.validateId(id, false)
+        .then(() => this.loadLayouts())
+        .then(existingLayouts => {
+          const layouts = existingLayouts.filter(layout => layout.id === id);
+          resolve(layouts[0].json);
+        })
+        .catch(e => reject(e));
+    });
   }
 
-  loadMetadata(): LayoutMetadata[] {
-    return getLocalEntity<LayoutMetadata[]>(metadataSaveLocation) || [];
+  loadMetadata(): Promise<LayoutMetadata[]> {
+    return new Promise((resolve) => {
+      const metadata = getLocalEntity<LayoutMetadata[]>(metadataSaveLocation);
+      resolve(metadata || []);
+    })
   }
 
-  private loadLayouts(): Layout[] {
-    return getLocalEntity<Layout[]>(layoutsSaveLocation) || [];
+  private loadLayouts(): Promise<Layout[]> {
+    return new Promise((resolve) => {
+      const layouts = getLocalEntity<Layout[]>(layoutsSaveLocation);
+      resolve(layouts || []);
+    });
   }
 
-  private appendAndPersist(newId: string,
-                           newMetadata: Omit<LayoutMetadata, "id">,
-                           newLayout: LayoutJSON,
-                           existingLayouts: Layout[],
-                           existingMetadata: LayoutMetadata[]) {
-    existingLayouts.push({id: newId, json: newLayout});
-    existingMetadata.push({id: newId, ...newMetadata});
+  private appendAndPersist(
+    newId: string,
+    newMetadata: Omit<LayoutMetadata, "id">,
+    newLayout: LayoutJSON,
+    existingLayouts: Layout[],
+    existingMetadata: LayoutMetadata[]
+  ) {
+    existingLayouts.push({ id: newId, json: newLayout });
+    existingMetadata.push({ id: newId, ...newMetadata });
 
     this.saveLayoutsWithMetadata(existingLayouts, existingMetadata);
   }
 
-  private saveLayoutsWithMetadata(layouts: Layout[], metadata: LayoutMetadata[]): void {
+  private saveLayoutsWithMetadata(
+    layouts: Layout[],
+    metadata: LayoutMetadata[]
+  ): void {
     saveLocalEntity<Layout[]>(layoutsSaveLocation, layouts);
     saveLocalEntity<LayoutMetadata[]>(metadataSaveLocation, metadata);
   }
+
+  private async validateIds(id: string): Promise<void> {
+    return Promise
+      .all([
+        this.validateId(id, true).catch(error => error.message),
+        this.validateId(id, false).catch(error => error.message)
+      ])
+      .then((errorMessages: string[]) => {
+        const combinedMessage = errorMessages.filter(Boolean).join("; ");
```

Comment:
const combinedMessage = errorMessages.reduce((acc, cur) => cur ? `${acc}; ${cur}` : acc, "")

Reply:
This would have to be modified slightly to avoid introducing a semicolon at the start of the combined message:
```javascript
errorMessages.reduce((acc, curr) => curr && acc ? `${acc}; ${curr}` : acc || curr, "")
```

Personally I prefer the filter/join approach (maybe with a comment to explain the filter), because at least the join part is intuitive. I think the reducer's callback function is more confusing with all the added truthiness. I'm still not used to seeing reducers though, so I'm happy to go with concensus if you guys still prefer it that way.

Reply:
I think the initial value can be omitted in this case reduce((acc, curr) => curr && acc ? `${acc}; ${curr}` : acc || curr) which makes it a bit more concise. Reduce() is used to produce a new value out of an array of values which is what we are trying to achieve here so I would prefer reduce() here over filter().join()

Reply:
`join` is also used to produce a new value out of an array. I've added a comment to explain the filter (4a855d46) - @cfisher-scottlogic to provide the deciding vote!

Reply:
that is true. you could actually think of that reduce as a join() with the ability to conditionally exclude some values

Reply:
I'm gonna have to go with `filter(Boolean).join()` (with the comment explaining what it's doing). It's more concise, and with the `.join()` I can see exactly what this line of code is meant to be doing. With reduce, although I've been able to figure it out, it's not immediately obvious to me.